rpcs3/Emu/Cell/Modules/cellSync2.cpp

```diff
@@ -135,56 +221,98 @@ s32 _cellSync2SemaphoreAttributeInitialize(vm::ptr<CellSync2SemaphoreAttribute>
 	return CELL_OK;
 }
 
-s32 cellSync2SemaphoreEstimateBufferSize(vm::cptr<CellSync2SemaphoreAttribute> attr, vm::ptr<u32> bufferSize)
+error_code cellSync2SemaphoreEstimateBufferSize(vm::cptr<CellSync2SemaphoreAttribute> attr, vm::ptr<u32> bufferSize)
 {
 	cellSync2.todo("cellSync2SemaphoreEstimateBufferSize(attr=*0x%x, bufferSize=*0x%x)", attr, bufferSize);
 
+	if (!attr || !bufferSize)
+		return CELL_SYNC2_ERROR_NULL_POINTER;
+
 	if (attr->maxWaiters == 0 || attr->maxWaiters > 32768)
 		return CELL_SYNC2_ERROR_INVAL;
 
 	return CELL_OK;
 }
 
-s32 cellSync2SemaphoreInitialize()
+error_code cellSync2SemaphoreInitialize(vm::ptr<CellSync2Semaphore> semaphore, vm::ptr<void> buffer, s32 initialValue, vm::cptr<CellSync2SemaphoreAttribute> attr)
 {
-	UNIMPLEMENTED_FUNC(cellSync2);
+	cellSync2.todo("cellSync2SemaphoreInitialize(semaphore=*0x%x, buffer=*0x%x, initialValue=0x%x, attr=*0x%x)", semaphore, buffer, initialValue, attr);
+
+	if (!semaphore || !attr || ((attr->maxWaiters >= 2) && !buffer))
+		return CELL_SYNC2_ERROR_NULL_POINTER;
+
+	if ((initialValue > 8388607) || (initialValue < -8388608) || (attr->maxWaiters == 0) || ((attr->maxWaiters == 1) && buffer))
+		return CELL_SYNC2_ERROR_INVAL;
+
 	return CELL_OK;
 }
 
-s32 cellSync2SemaphoreFinalize()
+error_code cellSync2SemaphoreFinalize(vm::ptr<CellSync2Semaphore> semaphore)
 {
-	UNIMPLEMENTED_FUNC(cellSync2);
+	cellSync2.todo("cellSync2SemaphoreFinalize(semaphore=*0x%x)", semaphore);
+
+	if (!semaphore)
+		return CELL_SYNC2_ERROR_NULL_POINTER;
+
 	return CELL_OK;
 }
 
-s32 cellSync2SemaphoreAcquire()
+error_code cellSync2SemaphoreAcquire(vm::ptr<CellSync2Semaphore> semaphore, u32 count, vm::cptr<CellSync2ThreadConfig> config)
 {
-	UNIMPLEMENTED_FUNC(cellSync2);
+	cellSync2.todo("cellSync2SemaphoreAcquire(semaphore=*0x%x, count=0x%x, config=*0x%x)", semaphore, count, config);
+
+	if (!semaphore)
+		return CELL_SYNC2_ERROR_NULL_POINTER;
+
+	if (count > 8388607)
```

Comment:
is there a name for this number in docs or can you make one :P? its just too many magic numbers imo

Reply:
it's 0x7FFFFF

Reply:
oh in hex it looks much better huh

Reply:
I replaced the "magic" numbers by their hexadecimal representation